llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+
+				if (isValueComputedConstant(v) == false)
+				{
+					if (isa<Argument>(v) && exe && exe->isValueComputed(v))
+					{
+
+					auto	     StoredAddr = (char*) exe->ValueAddresses->toReal(exe->getOperandValue(v).PointerVal);
+        const GlobalValue* GV = exe->getGlobalValueAtAddress(StoredAddr);
+
+	if (GV)
+	{
+	v = GV;
+	}
+	else
+	{
+		GenericValue value = exe->getOperandValue(v);
+//		value.print("uela\t");
+
+		v=nullptr;
+	}
+					}
+					else
+						v = nullptr;
+				}
+			}
+		}
+
+		return args;
+	}
+	static bool areEquivalent(const VectorOfArgs& a, const VectorOfArgs& b)
+	{
+		assert(a.size() == b.size());
+
+		const uint32_t numElements = a.size();
+	
+		for (uint32_t i=0; i<numElements; i++)
+		{
+			if (b[i] == nullptr)
+				continue;
+
+			if (a[i] != b[i])
+				return false;
+		}
+		return true;
+	}
+	void enqueCallEquivalent(const VectorOfArgs& arguments)
+	{
+		//Insert the arguments in the map
+
+		//Check wether we already visited something similar
+		for (const auto& args : callEquivalentQueue)
+		{
+			if (areEquivalent(arguments, args))
+			{
+				return;
+			}
+		}
+		
+		callEquivalentQueue.push_back(arguments);
+	}
+	const VectorOfArgs* getSomethingToVisit()
+	{
+		if (indexCallEquivalentQueue >= callEquivalentQueue.size())
+			return nullptr;
+		else
+			return &callEquivalentQueue[indexCallEquivalentQueue++];
+	}	
+	void visitCallEquivalent(const VectorOfArgs& arguments)
+	{
+		//Insert the arguments in the map
+		bool areAllArgsNullptr = true;
+		{
+			moduleData.setUpPartialInterpreter(F, currentEE);
+
+
+			assert(arguments.size() == F.getFunctionType()->getNumParams());
+
+			for (uint32_t i=0; i<arguments.size(); i++)
+			{
+				const llvm::Value* x = arguments[i];
+				if (!x)
+					continue;
+				areAllArgsNullptr = false;
+				llvm::Argument* ith_arg = F.getArg(i);
+				currentEE->assignToMaps(ith_arg, x);
+			}
+		//Do the visit
+		actualVisit(/*skipVisit*/areAllArgsNullptr);
+		
+		//Cleanup
+		doneVisitCallBase();
+		}
+
+	}
+	void actualVisit(bool skipVisit);
+	void visitCallBase(const llvm::CallBase* callBase)
+	{
+		const auto& equivalentArgs = getArguments(callBase, nullptr, nullptr);
+
+		enqueCallEquivalent(equivalentArgs);
+  		// Handle varargs arguments...
+	}
+	void enqueVisitNoInfo()
+	{
+		const auto& equivalentArgs = getArguments(nullptr, nullptr, nullptr);
+
+		enqueCallEquivalent(equivalentArgs);
+	}
+	void doneVisitCallBase()
+	{
+		bool removed = currentEE->removeModule(moduleData.getModulePtr());
+		(void)removed;
+		assert(removed);
+		delete currentEE;
+		currentEE = nullptr;
+	}
+	void cleanupBB(Function& function)
+	{
+		int BBs =0;
+		std::set<std::pair<llvm::BasicBlock*, llvm::BasicBlock*>> Edges;
+		for (auto& x : function)
+		{
+			BBs++;
+			for (auto* p : predecessors(&x))
+				Edges.insert({p, &x});
+		}
+
+		for (auto& p : Edges)
+
+		{
+			if (visitedEdges.count(p) == 0)
+				removeEdgeBetweenBlocks(p.first, p.second);
```

Comment:
Maybe it would make more sense to visit the CFG and directly removed all edges which are not part of the set of visited edges.

Reply:
I reworked the iteration integrating with other parts, but I think it's cleaner if the call to removeEdgeBetweenBlocks are done explicitly